### PR TITLE
prov/efa: Merge rxr_fabric into efa_fabric.

### PR DESCRIPTION
**Pull request by richwelchaws**
Currently, the code for the main operations for efa_fabric and rxr_fabric
is separate even after some underlyling functionality had been merged. This
introduces unnecesary complexity and maintenance requirements.

This patch addresses that by merging the rxr_fabric structure into the
efa_fabric structure and collapses rxr_fabric() and rxr_fabric_close()
into the corresponding functions for efa_fabric, namely efa_fabric() and
efa_fabric_close().

This change will reduce code complexity and allow for easier maintence going
forward.

Signed-off-by: Rich Welch <rlwelch@amazon.com>

### CODE REVIEW

File: prov/efa/src/efa_fabric.c [hunk 1]
```diff
@@ -927,6 +933,17 @@ static int efa_fabric_close(fid_t fid)
 	ret = ofi_fabric_close(&fab->util_fabric);
 	if (ret)
 		return ret;
+
+	if (rxr_env.enable_shm_transfer) {
```
[the remark this answers is not in the document]
Comment (richwelchaws) [reply prov/efa/src/efa_fabric.c:941]: I've renamed `fab` to `efa_fabric`. I'm going to submit the other change as part of a follow-on change.

File: prov/efa/src/efa_fabric.c [hunk 2]
```diff
@@ -927,6 +933,17 @@ static int efa_fabric_close(fid_t fid)
...
+		ret = fi_close(&fab->shm_fabric->fid);
+		if (ret)
+			return ret;
```
[the remark this answers is not in the document]
Comment (richwelchaws) [reply prov/efa/src/efa_fabric.c:947]: Done.

File: prov/efa/src/efa_fabric.c [hunk 3]
```diff
@@ -942,7 +959,10 @@ static struct fi_ops efa_fi_ops = {
 
 static struct fi_ops_fabric efa_ops_fabric = {
 	.size = sizeof(struct fi_ops_fabric),
-	.domain = efa_domain_open,
+	//Until EFA and RXR domains are fully merged, we use rxr_domain_open() as
+	//the entry point for both and it calls efa_domain_open() as needed. That
+	//is the reason we use rxr_domain_open() for the domain value below.
```
[the remark this answers is not in the document]
Comment (richwelchaws) [reply prov/efa/src/efa_fabric.c]: Thanks, changes made.

File: prov/efa/src/efa_fabric.c [hunk 4]
```diff
@@ -966,18 +986,50 @@ int efa_fabric(struct fi_fabric_attr *attr, struct fid_fabric **fabric_fid,
...
-		return ret;
+
+	if (ret)
+		goto err_free_fabric;
+
+	/* Open shm provider's fabric domain */
+	if (rxr_env.enable_shm_transfer) {
+		assert(!strcmp(shm_info->fabric_attr->name, "shm"));
+		ret = fi_fabric(shm_info->fabric_attr,
+				       &fab->shm_fabric, context);
```
[the remark this answers is not in the document]
Comment (richwelchaws) [reply prov/efa/src/efa_fabric.c]: As above, I made the rename change and fixed the indentation.

**Comment by wzamazon** [comment]
bot:aws:retest

**Review by wzamazon — APPROVED** [review]

**Review by rwespetal — APPROVED** [review]

**Pull request merged**